Add overview page for docker-compose to k8s migration (#27270)
kubernetes/website: content/en/docs/tutorials/compose-migration/_index.md

---
title: Migrating from compose workflows to Kubernetes
weight: 60
---

## Disclaimer

The [compose specification](https://github.com/compose-spec/compose-spec) is evolving, this documentation aims to cover the latest version of this compose specification.
We will assume that we are talking about the latest compose specifications which the v3 at this time of writing.

> If your compose files at not at the latest version specification, we suggest you try to upgrade to the latest spec version first. 

Reviewer: ```suggestion
If your compose files are not at the latest version specification, we suggest you try to upgrade to the latest spec version first. 
```

## Differences between compose specification and Kubernetes specification

> A set of workflows defined using the compose specification cannot be automated translated to a unique set of Kubernetes objects.

Reviewer: ```suggestion
A set of workflows defined using the compose specification cannot be automatically translated to a unique set of Kubernetes objects.
```

While some notions are very similar, there is not a single way to translate concepts from compose to Kubernetes.
There is no direct equivalence for many concepts of compose in Kubernetes.

Also, you have features that are available in the compose specifications that do not have a clear equivalent in Kubernetes.

Therefore, these content are only there to be suggestions and not a definitive and absolute guide.

> There is an [ongoing work](https://github.com/compose-spec/compose-spec/blob/wip_kube_mapping/KUBERNETES_MAPPING.md) on the compose specification side to give some information about compose to Kubernetes mapping.

Reviewer: ```suggestion
There is a [Compose - Kubernetes mapping](https://github.com/compose-spec/compose-spec/blob/wip_kube_mapping/KUBERNETES_MAPPING.md) on the compose specification side to give some information about compose to Kubernetes mapping.
```
Perhaps it's better to just list the name of that page than to say "ongoing work", since the page hasn't been updated for about two years.

## Stateful and stateless workloads

> One of the first step to migrate from compose to Kubernetes is to analyze among the different components of the compose files the sub-systems that are stateless, and the one that are stateful.

Reviewer: ```suggestion
One of the first steps to migrate from compose to Kubernetes is to analyze the different components of the compose files sub-systems that are stateless and stateful.
```
I hope I got the spirit of the sentence right above. Not sure it's quite right.

### Stateless applications

> Stateless applications is one which depends on no persistent storage.

Reviewer: ```suggestion
A stateless application is one which depends on no persistent storage.
```

The only thing your cluster is responsible for is the code, and other static content, being hosted on it.

> That's it, no changing databases, no writes and no left over files when the pod is deleted.

Reviewer: ```suggestion
That's it. The stateless application has no database changes, no writes, and no leftover files when the pod is deleted.
```
Check if that is what you meant.

---
title: Migrating from compose workflows to Kubernetes
weight: 60
---

## Disclaimer

> The [compose specification](https://github.com/compose-spec/compose-spec) is evolving, this documentation aims to cover the latest version of this compose specification.

Reviewer: The version of the compose-spec seems unclear. The reference implementation seems to be [Docker Compose v2](https://github.com/docker/compose). It would be good to clarify this.

We will assume that we are talking about the latest compose specifications.
If your compose files at not at the latest version specification, we suggest you try to upgrade to the latest spec version first. 

## Differences between compose specification and Kubernetes specification

> A set of workflows defined using the compose specification cannot be automated translated to a unique set of Kubernetes objects.

Reviewer 2: ```suggestion
A set of workflows defined using the compose specification cannot be automatically translated to a unique set of Kubernetes objects.
```

This suggestion is a guess, but "automated translated" doesn't make any sense.

While some notions are very similar, there is not a single way to translate concepts from compose to Kubernetes.
For instance, a [service](https://github.com/compose-spec/compose-spec/blob/master/spec.md#services-top-level-element) concept in compose can be translated in different way in Kubernetes.

> In particular a service in compose **IS NOT** equivalent to a service in Kubernetes.

Reviewer 2: This paragraph is very confusing. Could you just say "There is no direct equivalence for many concepts in compose in Kubernetes" instead?

> Also, you have features that are available in the compose specifications that do not have a clear equivalent in Kubernetes.

Reviewer 2: You say this above, so I don't think you need to say it again. 

```suggestion
```

> Therefore, these content are only there to be suggestions and not a definitive and absolute guide.

Reviewer 2: ```suggestion
As such, this guide is a set of suggestions rather than a definitive guide.
```

There is an [ongoing work](https://github.com/compose-spec/compose-spec/blob/wip_kube_mapping/KUBERNETES_MAPPING.md) on the compose specification side to give some information about compose to Kubernetes mapping.

## Stateful and stateless workloads

> One of the first step to migrate from compose to Kubernetes is to analyze among the different components of the compose files the sub-systems that are stateless, and the one that are stateful.

Reviewer 2: ```suggestion
The first step when migrating from compose to Kubernetes is analyzing the different components of the compose files for sub-systems that are stateless, and sub-systems that are stateful.
```

### Stateless applications

> Stateless applications is one which depends on no persistent storage.
> The only thing your cluster is responsible for is the code, and other static content, being hosted on it.
> That's it, no changing databases, no writes and no left over files when the pod is deleted.

Reviewer 2: ```suggestion
Stateless applications do not depend persistent storage.
The only thing your cluster is responsible for is the code and other static content being hosted on it: no changing databases, no writes and no leftover files when the pod is deleted.
```

Stateless components are often migrated as [Deployment](/docs/concepts/workloads/controllers/deployment/) objects in Kubernetes.

### Stateful applications

> Stateful applications save data to persistent disk storage for use by the server, by clients, and by other applications.
> An example of a stateful application is a database or key-value store to which data is saved and retrieved by other applications.
> In the case of compose, this data could be persisted either on disk or on a different host.

Reviewer 2: ```suggestion
Stateful applications save data to persistent disk storage for use by the server, by clients, and by other applications.
An example of a stateful application is a database or key-value store to which data is saved and retrieved by other applications.
When using compose, this data is either persisted on disk or on a different host.
```
